Doc/howto/concurrency.rst

```diff
@@ -0,0 +1,617 @@
+and sometimes multiple execution pipelines per core.  The operating
+system exposes physical CPU threads as OS threads and as processes.
+A programming language (or runtime) may add additional layers of
+abstraction on top of that.
+
+Parallelism is where concurrent logical threads are running
+on distinct physical threads across multiple cores,
+
+Concurrency Models
+------------------
+
+The concept of concurrency has been a part of the study and practice
+of computer software since very early on, in the 1950s and 1960s,
+long before the wide-spread adotion of multi-core CPUs.  Clearly
+its about more than just parallelism.
+
+Over the decades, research and use of concurrency has led to a variety
+of well defined abstract models, with different characteristics and
+tradeoffs.  The application of the different theoretical concurrency
+models can be categorized as follows:
+
+================= ==========
+free threads      using multiple physical threads in the same process,
+                  with no isolation between them
+isolated threads  threads, often physical, with strict isolation
+                  between them (e.g. CSP and actor model)
+multiprocessing   using multiple isolated processes
+distributed       multiprocessing across multiple computers
+async/await       using coroutines (AKA "cooperative multitasking")
+================= ==========
+
+(There are certainly others, but these are the focus here.)
+
+There are tradeoffs to each.  Free-threading probably has the most
+notoriety and the most examples, but is also the most likely to cause
+you pain.
+Isolated threads have few of the downsides but are less familiar.
+Multiprocessing and distributed are less efficient at smaller scales.
+Async can be straightforward, but may cascade throughout a code base
+and doesn't necessarily give you parallelism.
+
+What problems can concurrency help solve?
+-----------------------------------------
+
+Primarily, concurrency can be helpful by making your program faster
+and more responsive (less latency), when possible.  In other words,
+you get better computational throughput.  That happens by enabling
+the following:
+
+* run on multiple CPU cores (parallelism)
+* keep blocking resources from blocking the whole program
+* make sure critical tasks have priority
+* make sure other tasks have a fair share of time
+* process results as they come, instead of waiting for them all
+
+Other possible benefits:
+
+* asynchronous events can be handled more cleanly
+* better efficiency using hardware resources
+* improved scalability
+
+What are the downsides?
+-----------------------
+
+The main challenge when using concurrency is the (potential) extra
+complexity.  This complexity comes from the effect of multiple logical
+threads running at the same time and interacting with each other.
+In practice, this falls into two categories: data races and tracing
+relative execution.  Both are a form of "spooky action at a distance".
```

Comment:
I've never heard this term before? Is this an analogy?

Reply:
Albert Einstein originally said it about quantum entanglement, [^1]  I suppose I could give more context (or drop the phrase). 😄

[^1]: https://en.wikipedia.org/wiki/Action_at_a_distance#%22Spooky_action_at_a_distance%22

Reply:
done